Config: Separate from constants

sentientwaffle opened the pull request:
## Background

Distinguish "config" from "constants":

- Configuration is anything that might be modified at build-time.
- Constants include:
  - Configuration (Constants are a superset of config).
  - Constants derived from configuration.
  - Universal constants (e.g. `sector_size`).

Distinguish between "process config" and "cluster config":
- _process config_ can be modified per-replica/client — different process do not need to agree, and a process can be safely restarted with a distinct process config.
- _cluster config_ must be identical on call replicas/clients — it factors in to the storage format and/or network protocol.

## Changes

- Rename `config.zig` to `constants.zig` (but we still use `const config = @import("constants.zig");` — I didn't want to make the diff even more unmanageable).
- A preset config can be chosen at build time. By default:
  - `main.zig` builds with the `development` config
  - Tests, fuzzers, and the VOPR build with the `test_min` config.
- Groove/Tree: Allow zero-sized value caches (https://github.com/tigerbeetledb/tigerbeetle/pull/301/commits/c53647432fe25fdbad7b440e71d54b705560d50b).
- Change the default value cache sizes: favor accounts, don't cache transfers.

There are currently 3 preset configs:

- `test_min` is a minimal configuration to allow deep testing of the LSM in-memory.
- `production`
- `development` is `production` with smaller caches (and a compatible storage format).

## Future work
(For a future PR since this one is functional but pretty giant already...)

- Rename `config` to `constants` everywhere.
- Support more granular config updates at build time (e.g. `-Dconfig-log-level=debug`).
- Update `tigerbeetle version --verbose` — it should probably only print config (not constants), and distinguish between "process" configs and "cluster" configs.
- https://github.com/tigerbeetledb/tigerbeetle/issues/295
- Fuzz/VOPR _multiple_ different configs instead of always using `test_min`.

sentientwaffle committed:
Config: Remove unused configs

sentientwaffle committed:
Config: Distinguish config from constants

There are (roughly) three categories of constants in `config.zig`:
- `ConfigCluster` constants can be modified when the cluster is first configured, but must never change after that.
- `ConfigProcess` can vary per-replica/client, and can be changed between restarts.
- And various constants derived from the above (plus a few that just have fixed values, like `sector_size`).

I strongly considered threading a `constants` type through all of the type constructors from `main.zig`/`simulator.zig` on down. But this adds quite a lot of "noise" to the code, since there are many functions/constants/types that aren't generic already.

The main goal is to allow the VOPR and other fuzzers to test different configurations — this is still achievable with to "global" `config.zig`.

The downsides of the "global" `config.zig`:
- Only one config can be run per process. e.g. the VOPR cannot test multiple configs simultaneously.
- If a unit test needs to target a specific configuration, it needs to specifically factor out the config. (See `SuperBlockQuorums` for an example).

sentientwaffle committed:
Config: Remove config.Environment

sentientwaffle committed:
Config: Run test with config_test_min

sentientwaffle committed:
Config: Minimize test config

sentientwaffle committed:
LSM: Allow zero-sized value cache

This also moves ownership of the value cache from the groove(s) to the tree, just to clean things up.

sentientwaffle committed:
Config: Disable transfer value cache by default

Also increase the sizer of the account cache, relative to the posted-transfer cache. (For reasons explained in `config.zig`).

sentientwaffle committed:
Config: Get base config from "zig build"

By default:
- When running tests, use `test_min`.
- For fuzzers, use `test_min`.
- Otherwise use `development`.

sentientwaffle committed:
Config: Rename io_depth_* → journal_iops_*_max

These config optional only apply to the journal. The grid's IOPS limits are presently hard-coded (see https://github.com/tigerbeetledb/tigerbeetle/issues/295).

sentientwaffle committed:
Config: Manually export "log"

To match `log_level`.

sentientwaffle committed:
Config: Simplify build options

sentientwaffle committed:
Config: Spread out assertions

Config constraints are borderline documentation, so keep them close to their respective declarations.

sentientwaffle committed:
Config: Rename to constants

This only renames the file — it is still referred to `config`.
(I don't want to trigger a zillion git conflicts...)

jamii commented on build.zig:
```diff
@@ -24,12 +24,18 @@ pub fn build(b: *std.build.Builder) void {
+
     const tracer_backend = b.option(
-        TracerBackend,
+        config.TracerBackend,
```
Now that TracerBackend is in config, we can use it in https://github.com/tigerbeetledb/tigerbeetle/blob/dec1160d09494d9853a166153b953b9f6d6cdc59//src/clients/node/src/node.zig#L32 too.

jamii commented on src/config.zig:
```diff
@@ -1,413 +1,198 @@
+    rtt_multiple: u8 = 2,
+    backoff_min_ms: u64 = 100,
+    backoff_max_ms: u64 = 10000,
+    clock_offset_tolerance_max_ms: u64 = 10000,
```
I haven't looked at how the clock sync works yet, but I'm suprised that different replicas can have different settings for it.

sentientwaffle replied:
You probably _shouldn't_ use different values, but we don't rely on clock sync for correctness, only availability. `ConfigProcess` is for things that are safe to change at "cluster runtime" — e.g. you could do a rolling replica restart to deploy a new value.

jamii replied:
:+1:

jamii commented:
Looks good.

sentientwaffle commented:
Oops, I need to fix the node client

jamii commented:
Have you tried running tree_fuzz yet with the min config? I'd be curious to see how many lsm levels it can reach now.

sentientwaffle committed:
Config: Fix node client tracer backend

sentientwaffle commented:
I did! It found https://github.com/tigerbeetledb/tigerbeetle/pull/296 and https://github.com/tigerbeetledb/tigerbeetle/pull/300.

In the run I was debugging yesterday it reached level 4.

sentientwaffle commented:
I'm trying to figure out why the go client tests fail, then this will be good to go.

jamii commented:
I need to learn to refresh the page to avoid stale CI results :S

sentientwaffle committed:
CLI: Allow cache=0

sentientwaffle committed:
Fuzz,CI: Add --events-max

Allow a runtime argument to limit the number of fuzzer repetitions.

CI canary runs can be fast, but for actual fuzzing we can still use a high default events-max.

(The motivation for this was a (successful) ManifestLog fuzzer run that took >5 hours.)

sentientwaffle commented:
Ignore this for a while; I botched a rebase...

jamii commented on .github/workflows/test_and_build.yml:
```diff
@@ -100,10 +100,10 @@ jobs:
     steps:
       - uses: actions/checkout@v2
       - run: ./scripts/install_zig.sh
-      - run: zig/zig build fuzz_vsr_superblock -- --seed 123
```
Should we put tree/forest fuzz in CI too?

sentientwaffle replied:
I'll add the tree fuzzer in a follow-up PR. There is a bug in the forest fuzzer (revealed by the new minimal config) that makes it fail right now though.
